docs/samediff/templates/ops.md

```diff
@@ -1,3 +1,243 @@
+for example, if we want to create a 3x3 convolutional layer with `relu` activation, we may proceed as follows:
+```java
+Conv2DConfig config2d = new Conv2DConfig()
+    .kW(3)
+    .kH(3)
+    .pW(2)
+    .pH(2);
+SDVariable convolution2dLinear = sameDiff.cnn.conv2d(input, weights, config2d);
+SDVariable convolution2dOutput = sameDiff.nn.relu(convolution2dLinear);
+``` 
+In the first line, we construct a convolution configuration using its default constructor. Then we specify the 
+kernel size (this is mandatory) and optional padding size, keeping other settings default (unit stride, no 
+dilation, no bias, `NCHW` data format). We then employ this configuration to create a linear convolution with predefined
+`SDVariables` for input and weights; the shape of `weights` is to be tuned to that of `input` and to `config` 
+beforehand. Thus, if in the above example `input` has shape, say, `[-1, nIn, height, width]`, then `weights` are to have
+a form `[nIn, nOut, 3, 3]` (because we have 3x3 convolution kernel). The shape of the resulting variable `convoluton2d` 
+will be predetermined by these parameters (in our case, it will be `[-1, nOut, height, width]`). Finally, in the last 
+line we apply a `relu` activation.
+
+### `rnn` - Recurrent neural networks
+
+This module contains arguably the most sophisticated methods in the framework. Currently it allows you to create 
+- simple recurrent units, using `sru` and `sruCell` methods;
+- LSTM units, using `lstmCell`, `lstmBlockCell` and `lstmLayer`;
+- Graves LSTM units, using `gru` methods.
+
+Like convolution-type operations, recurrent operations require special configuration objects as input. Now, however, you 
+actually need to pack all the variables that will be used in a unit into the configuration object. For instance, to 
+create a simple recurrent unit, you need to proceed like that:
+```java
+SRUConfiguration sruConfig = new SRUConfiguration(input, weights, bias, init);
+SDVariable sruOutput = samediff.rnn().sru(sruConfig);
+```
+Here, the arguments in the `SRUConfiguration` constructor are variables that are to be defined beforehand. Obviously 
+their shapes should be matching, and these shapes predetermine the shape of `output`.
+
+### `loss` - Loss functions
+In this branch we keep common loss functions. Most loss functions may be created quite simply, like that:
+```java
+SDVariable logLoss = sameDiff.loss.logLoss("logLoss", label, predictions);
+```
+where `labels` and `predictions` are `SDVariable`'s. A `String` name is a mandatory parameter in most `loss` methods, 
+yet it may be set to `null` - in this case, the name will be generated automatically. You may also create weighted loss
+functions by adding another `SDVariable` parameters containing weights, as well as specify a reduction method for the 
+loss over the minibatch. Thus, a full-fledged `logLoss` operation may 

```

Comment:
"reduction method" - let's basically copy paste the content from here to explain the reduction modes:
https://github.com/eclipse/deeplearning4j/blob/master/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/loss/LossReduce.java